osquery/tables/system/windows/shimcache.cpp

```diff
@@ -0,0 +1,216 @@
+    LOG(WARNING) << "ERROR: Failed to decode Shimcache hex values to string";
+    shimcache.last_modified = 0LL;
+    return shimcache;
+  }
+
+  shimcache.path = string_path;
+  int shimcache_modified_start = 0;
+  int execution_flag_start = 0;
+
+  // If execution flag exists set where the flag starts in the substring
+  if (execution_flag_exists == "true") {
+    shimcache_modified_start = 40;
+    execution_flag_start = 24;
+  } else {
+    shimcache_modified_start = 20;
+  }
+  std::string shimcache_time =
+      token.substr(shimcache_modified_start + shimcache_file_path * 2, 16);
+
+  // Check to make sure the shimcache entry is greater than zero
+  if (string_path.length() > 0) {
+    // Get last modified time of the file
+    shimcache.last_modified = lastExecute(shimcache_time);
+  } else {
+    shimcache.last_modified = 0LL;
+  }
+
+  if (execution_flag_exists == "true") {
+    int shimcache_flag = std::stoi(
+        token.substr(execution_flag_start + shimcache_file_path * 2, 2),
+        nullptr,
+        16);
+    // Perform Bitwise AND to determine TRUE or FALSE
+    if (shimcache_flag & 2) {
+      shimcache.execution_flag = "true";
+    } else {
+      shimcache.execution_flag = "false";
+    }
+  }
+  return shimcache;
+}
+
+auto parseRegistry(std::set<std::string> shimcacheControl,
+                   Row r,
+                   QueryData results) {
+  for (const auto& rKey : shimcacheControl) {
+    std::size_t shimcache_entry = rKey.find(
+        "Control\\Session "
+        "Manager\\AppCompatCache");
+
+    if (shimcache_entry != std::string::npos) {
+      QueryData shimcacheResults;
+      int entry = 1;
+
+      queryKey(rKey, shimcacheResults);
+      for (const auto& aKey : shimcacheResults) {
+        if (aKey.at("name") == "AppCompatCache") {
+          std::string token;
+          std::string execution_flag_exists;
+          std::string delimter;
+
+          size_t pos = 0;
+          std::string data = aKey.at("data");
+
+          // Check if Registry data starts with any of supported WIN_START
+          // values and if the Shimcache delimiter exists at the specific
+          // substring
+          if ((boost::starts_with(data, WIN8_START)) &&
+              (data.substr(WIN8, 8) == WIN_81_10_SHIMCACHE_DELIMITER)) {
+            execution_flag_exists = "true";
+            delimter = WIN_81_10_SHIMCACHE_DELIMITER;
+          } else if (boost::starts_with(data, WIN10_START) &&
+                     (data.substr(WIN10_PRE_CREATOR, 8) ==
+                      WIN_81_10_SHIMCACHE_DELIMITER)) {
+            delimter = WIN_81_10_SHIMCACHE_DELIMITER;
+          } else if (boost::starts_with(data, WIN10_CREATOR_START) &&
+                     (data.substr(WIN10_CREATOR, 8) ==
+                      WIN_81_10_SHIMCACHE_DELIMITER)) {
+            delimter = WIN_81_10_SHIMCACHE_DELIMITER;
+          } else {
+            LOG(WARNING) << "Unknown or unsupported shimcache data!";
```

Comment:
Prefer to have only one `LOG(WARNING)` per event, this makes it easy to count and understand warnings and errors when collected at scale.

```
LOG(WARNING) << "Unknown or unsupported shimcache data: " << data.substr(256, 8);
```

and remove the following line.

Reply:
Fixed